src/sparse/vec.rs

```diff
@@ -486,27 +486,150 @@ where
+            if i.to_usize().unwrap() >= n {
+                return Err((
+                    indices,
+                    data,
+                    StructureError::SizeMismatch(
+                        "indices larger than vector size",
+                    ),
+                ));
+            }
+        }
+        Ok(Self::new_trusted(n, indices, data))
+    }
+
+    /// Create a sparse vector
     ///
     /// # Panics
     ///
     /// - if `indices` and `data` lengths differ
     /// - if the vector contains out of bounds indices
-    pub fn new(n: usize, indices: Vec<I>, data: Vec<N>) -> Self
-    where
-        N: Copy,
-    {
-        Self::try_new(n, indices, data).unwrap()
+    /// - if indices are out of order
+    ///
+    /// # Examples
+    /// ```rust
+    /// # use sprs::*;
+    /// // Creating a sparse owned vector
+    /// let owned = CsVec::new(10, vec![0, 4], vec![-4, 2]);
+    /// // Creating a sparse borrowing vector with `I = u16`
+    /// let borrow = CsVecViewI::new(10, &[0_u16, 4], &[-4, 2]);
+    /// // Creating a general sparse vector with different storage types
+    /// let mixed = CsVecBase::new(10, &[0_u64, 4] as &[_], vec![-4, 2]);
+    /// ```
+    pub fn new(n: usize, indices: IStorage, data: DStorage) -> Self {
+        Self::try_new(n, indices, data)
+            .map_err(|(_, _, e)| e)
+            .unwrap()
     }
 
-    /// Try create an owning `CsVec` from vector data.
+    /// Try create a sparse vector
     pub fn try_new(
         n: usize,
-        indices: Vec<I>,
-        data: Vec<N>,
-    ) -> Result<Self, StructureError>
+        indices: IStorage,
+        data: DStorage,
+    ) -> Result<Self, (IStorage, DStorage, StructureError)> {
+        Self::new_(n, indices, data)
+    }
+
+    /// Internal version of `new_unchecked` where we guarantee the invariants
+    /// ourselves
+    pub(crate) fn new_trusted(
+        n: usize,
+        indices: IStorage,
+        data: DStorage,
+    ) -> Self {
+        Self {
+            dim: n,
+            indices,
+            data,
+        }
+    }
+
+    /// Create a `CsVec` without checking the structure
+    ///
+    /// # Safety
+    ///
+    /// This is unsafe because algorithms are free to assume
+    /// that properties guaranteed by [`check_structure`](CsVecBase::check_structure) are enforced.
+    /// For instance, non out-of-bounds indices can be relied upon to
+    /// perform unchecked slice access.
+    pub unsafe fn new_uncheked(
+        n: usize,
+        indices: IStorage,
+        data: DStorage,
+    ) -> Self {
+        Self {
+            dim: n,
+            indices,
+            data,
+        }
+    }
+}
+
+impl<N, I: SpIndex, DStorage, IStorage> CsVecBase<IStorage, DStorage, N, I>
+where
+    DStorage: std::ops::DerefMut<Target = [N]>,
+    IStorage: std::ops::DerefMut<Target = [I]>,
+{
+    /// Creates a sparse vector
+    ///
+    /// Will sort indices and data if necessary
+    pub fn new_sorted(
```

Comment:
Maybe it would make more sense naming this 'new_unsorted`?

Reply:
I think I prefer `new_sorted`, `new_unsorted` makes it sound like we support unsorted indices in the storage. Or maybe `new_from_unsorted` is the more explicit version?

Reply:
`new_from_unsorted` is very explicit in what it tries to do. I'll rename to this

Reply:
Should we rename the matrix version to also be `new_from_unsorted` to have a coherent API?

Reply:
Yes indeed!

Reply:
Renamed in aedc570